spec/index.bs

```diff
@@ -726,8 +714,49 @@ interface GPUBufferUsage {
 };
 </script>
 
+In order to guarantee immediate, efficient, and portable mapping, buffers
+marked for {{GPUBufferUsage/MAP_WRITE}} are only writable by the CPU,
+and cannot be also marked for {{GPUBufferUsage/COPY_DST}} or {{GPUBufferUsage/STORAGE}}.
+
 ## Buffer Mapping ## {#buffer-mapping}
 
+<script type=idl>
+typedef unsigned long GPUMapFlags;
+interface GPUMap {
+    const GPUMapFlags READ  = 0x0001;
+    const GPUMapFlags WRITE = 0x0002;
+};
+</script>
+
+{{GPUBuffer/map()}} returns null if the device is lost.
+
+Because MAP_WRITE is writable only by the CPU, the contents of the ArrayBuffer
+are unchanged between unmap and map. (Or, if the buffer has not been previously
+mapped, the initial buffer contents.
+
+If the contents of a mapping without {{GPUMap/WRITE}} are changed, on unmap(),
+the device is lost.
+Additionally, writing to a read-only ArrayBuffer returned from mapping without
+{{GPUMap/WRITE}} MAY throw an error.
+
+Only one mapping may be active per buffer at a time.
+
+With {{GPUQueue/createFence()}} and awaiting {{GPUFence/onCompletion()}}, it's possible
+to ensure that {{GPUBuffer/map()}} provides an efficient mapping to memory with a minimal
+number of copies.
+
+If applications do not ensure proper fencing, {{GPUBuffer/map()}} may provide a less
+efficient mapping. (such as requiring an extra allocation and copy) Implementations should
+provide warnings to inform when this overhead is incurred.
+
+{{GPUMap/READ}} will stall if its contents are stale. Because this condition is
```

Comment:
Stalling isn't an option in Web APIs anymore. WebGL got away with it but this is no longer a design option.

Reply:
Please include a reference to this rule.

Reply:
https://w3ctag.github.io/design-principles/#wrapper-apis

Reply:
Right, this blocking behavior is why the existing design returns promises. It's important to not block - we don't know how long the GPU will be using the resource. A long compute job could take seconds, and it's not okay to hang the main browser thread for seconds.

A better design would probably instead be for the map operation to fail outright.

Reply:
I didn't realize that you already made a pull request for fallible mapping: https://github.com/gpuweb/gpuweb/pull/511